hyper/http20/connection.py

```diff
@@ -98,6 +99,39 @@ def __init__(self, host, port=None, secure=None, window_manager=None, enable_pus
         #: Defaults to 64kB.
         self.network_buffer_size = 65536
 
+        # Concurrency
+        #
+        # Use one lock (_lock) to synchronize
+        # - connect()
+        #   ensures that socket creation, sending the preamble, and the initial
+        #   settings exchange occur just once.
+        # - _new_stream()
+        #   ensures that stream creation sees a consistent connection state
+        # - close()
+        #   ensures that closing threads see a consistent connection state
+        #
+        # It's ok to use the same in lock all these cases as they occur at
+        # different/linked points in the connection's lifecycle.
+        #
+        # Use another 2 locks (_write_lock, _read_lock) to synchronize
+        # - _send_cb
+        # - _recv_cb
+        # respectively
+        #
+        # I.e, sends/receve on the connection and its streams are serialized
+        # and separately across the threads accessing the connection.  This is
+        # a simple way of providing thread-safety.
+        #
+        # _write_lock and _read_lock synchronize all interactions between
+        # streams and the connnection.  There is a third i/o callback,
+        # _close_stream, passed to a stream's constructor.  It does not need to
+        # be synchronized, it uses _send_cb internally (which is serialized);
+        # it's other activity (safe deletion of the stream from self.streams)
+        # does not require synchronization.
```

Comment:
It might be worth updating the comment here to talk about the correct order to obtain these locks in the event that more than one is required. This is particularly true for hyper, which gets a bit carried away about calling back into itself recursively. ;)